File: crates/phoenix-codegen/tests/roundtrip/python/decimal_driver.py
# Behavioral Decimal round-trip driver for the Python target.
#
# Committed source. The Rust harness (`roundtrip.rs::decimal_python_roundtrip`)
# generates the small Decimal schema into ./generated_decimal/ (separate from the
# other round-trips' packages), then runs this with the committed `.venv`. It
# drives the generated httpx client in-process against the generated FastAPI
# server via `httpx.ASGITransport` and proves `Decimal` round-trips as exact
# decimal strings both ways: body decimals (required / optional / list / map) come
# back as equal `decimal.Decimal`s (proving `model_dump(mode="json")` → server
# parse → revalidate); a `Decimal` query param round-trips (echoed by the stub
# into the body); a required response-header decimal round-trips via
# `str()` / `Decimal(...)`; and a bare `Decimal` response decodes via
# `Decimal(response.json())`. pydantic validates every Decimal on parse, so the
# malformed body + query raw requests must be rejected. Exits non-zero on failure.

from __future__ import annotations

import asyncio
import sys
from decimal import Decimal

import httpx
from fastapi import FastAPI

from generated_decimal import models as m
from generated_decimal.client import ApiClient
from generated_decimal.server import create_router


def fail(msg: str) -> None:
    print(f"FAIL: {msg}")
    sys.exit(1)


COMPUTED_TAX = Decimal("8.25")


class Stub:
    async def echo_invoice(self, body: m.EchoInvoiceBody) -> m.Invoice:
        return m.Invoice(
            id=body.id,
            subtotal=body.subtotal,
            discount=body.discount,
            line_totals=body.line_totals,
            rates=body.rates,
        )

    async def get_quote(self, id: str, *, min_amount: Decimal) -> m.GetQuoteResult:
        return m.GetQuoteResult(
            body=m.Invoice(id=1, subtotal=min_amount, line_totals=[], rates={}),
            computed_tax=COMPUTED_TAX,
        )

    async def exchange_rate(self) -> Decimal:
        return Decimal("1.0825")


async def main() -> None:
    app = FastAPI()
    app.include_router(create_router(Stub()))
    client = ApiClient("http://test")
    client.client = httpx.AsyncClient(
        transport=httpx.ASGITransport(app=app), base_url="http://test"
    )

    subtotal = Decimal("19.99")
    discount = Decimal("-2.50")
    a = Decimal("10.00")
    b = Decimal("9.99")

    resp = await client.echo_invoice(
        m.EchoInvoiceBody(
            id=7,
            subtotal=subtotal,
            discount=discount,
            line_totals=[a, b],
            rates={"usd": Decimal("1.0"), "eur": Decimal("0.92")},
        )
    )
    if resp.subtotal != subtotal:
        fail(f"echo subtotal: {resp.subtotal} != {subtotal}")
    if resp.discount != discount:
        fail(f"echo discount: {resp.discount} != {discount}")
    if resp.line_totals != [a, b]:
        fail(f"echo line_totals: {resp.line_totals}")
    if resp.rates != {"usd": Decimal("1.0"), "eur": Decimal("0.92")}:
        fail(f"echo rates: {resp.rates}")

    min_amount = Decimal("5.00")
    r2 = await client.get_quote("inv-1", min_amount=min_amount)
    if r2.body.subtotal != min_amount:
        fail(f"query decimal not round-tripped: {r2.body.subtotal} != {min_amount}")
    if r2.computed_tax != COMPUTED_TAX:
        fail(f"computed_tax header: {r2.computed_tax}")

    rate = await client.exchange_rate()
    if rate != Decimal("1.0825"):
        fail(f"bare decimal response: {rate}")

    # Reject path: a malformed body decimal must be rejected by pydantic's Decimal
    # parse on the server. POST raw JSON to bypass the typed client (which would
    # refuse to construct the model locally), so the server-side decode fails.
    bad = await client.client.post(
        "/invoices",
        json={"id": 1, "subtotal": "not-a-number", "lineTotals": [], "rates": {}},
    )
    if bad.status_code < 400:
        fail(f"server accepted malformed body decimal: HTTP {bad.status_code}")

    # Reject path (query decimal): unlike Go, Python validates query decimals —
    # FastAPI coerces the `min_amount: Decimal` param and 422s on malformed input.
    bad_q = await client.client.get("/quote/inv-1", params={"minAmount": "not-a-number"})
    if bad_q.status_code < 400:
        fail(f"server accepted malformed query decimal: HTTP {bad_q.status_code}")

    print("OK")


asyncio.run(main())

File: crates/phoenix-codegen/tests/roundtrip/typescript/decimal-driver.ts
// Behavioral Decimal round-trip driver for the TypeScript target (Express).
//
// Committed source. The Rust harness (`roundtrip.rs::decimal_typescript_roundtrip`)
// generates the small Decimal schema into ./generated-decimal/ (separate from the
// other round-trips' generated dirs), then runs this via `tsx`. It proves the
// branded `Decimal` alias and the `parseDecimal` validate-on-decode pass
// round-trip exact decimal strings: body decimals (required / optional / list /
// map) come back equal and still typed `string` at runtime (the brand is
// compile-time only); a `Decimal` query param round-trips (echoed by the stub
// into the body); a required response-header decimal arrives; and a bare
// `Decimal` response decodes. It also asserts a malformed body decimal AND a
// malformed query decimal are rejected (TS validates both — the Go driver pins
// the opposite for the query). Exits nonzero on failure.

import express from "express";
import type { AddressInfo } from "node:net";

import { createRouter } from "./generated-decimal/server";
import type { Handlers } from "./generated-decimal/handlers";
import { api, setBaseUrl } from "./generated-decimal/client";
import { parseDecimal } from "./generated-decimal/types";
import type {
  Decimal,
  EchoInvoiceBody,
  GetQuoteResult,
  Invoice,
} from "./generated-decimal/types";

function check(cond: boolean, msg: string): void {
  if (!cond) {
    console.error(`FAIL: ${msg}`);
    process.exit(1);
  }
}

const computedTax = parseDecimal("8.25");

const stub: Handlers = {
  // Echo the decoded (server-side reviver-validated) body back unchanged.
  echoInvoice(body: EchoInvoiceBody): Promise<Invoice> {
    check(typeof body.subtotal === "string", "server: body.subtotal is a string");
    return Promise.resolve(body as Invoice);
  },
  // Echo the query decimal into `subtotal` so the client can assert it round-tripped.
  getQuote(id: string, query: { minAmount: Decimal }): Promise<GetQuoteResult> {
    return Promise.resolve({
      body: { id: 1, subtotal: query.minAmount, lineTotals: [], rates: {} },
      computedTax,
    });
  },
  exchangeRate(): Promise<Decimal> {
    return Promise.resolve(parseDecimal("1.0825"));
  },
};

async function main(): Promise<void> {
  const app = express();
  app.set("etag", false);
  app.use(express.json());
  app.use(createRouter(stub));
  const server = app.listen(0);
  await new Promise<void>((resolve) => server.once("listening", resolve));
  const port = (server.address() as AddressInfo).port;
  setBaseUrl(`http://127.0.0.1:${String(port)}`);

  const subtotal = parseDecimal("19.99");
  const discount = parseDecimal("-2.50");
  const a = parseDecimal("10.00");
  const b = parseDecimal("9.99");

  const resp = await api.echoInvoice({
    id: 7,
    subtotal,
    discount,
    lineTotals: [a, b],
    rates: { usd: parseDecimal("1.0"), eur: parseDecimal("0.92") },
  });
  check(typeof resp.subtotal === "string" && resp.subtotal === subtotal, "echo subtotal");
  check(resp.discount === discount, "echo optional discount");
  check(resp.lineTotals.length === 2 && resp.lineTotals[1] === b, "echo list line total");
  check(resp.rates.eur === parseDecimal("0.92"), "echo map rate");

  const minAmount = parseDecimal("5.00");
  const r2 = await api.getQuote("inv-1", { minAmount });
  check(r2.body.subtotal === minAmount, "query Decimal round-tripped through the server");
  check(r2.computedTax === computedTax, "response-header Decimal arrived");

  const rate = await api.exchangeRate();
  check(typeof rate === "string" && rate === parseDecimal("1.0825"), "bare Decimal response decoded");

  // Reject path: a malformed body decimal must fail the server's body reviver
  // (`parseDecimal` throws → 500 → client throws on `!response.ok`). The cast
  // smuggles a bad value past `parseDecimal` so the wire decode rejects it.
  let rejected = false;
  try {
    await api.echoInvoice({
      id: 1,
      subtotal: "not-a-number" as unknown as Decimal,
      lineTotals: [],
      rates: {},
    });
  } catch {
    rejected = true;
  }
  check(rejected, "server rejected malformed body decimal");

  // Reject path (query param): TS validates query/request-header `Decimal`s
  // inline via `parseDecimal` on the server, which throws `ValidationError` → 400
  // (like an enum param), matching Go's `decimalRe` check and Python's FastAPI
  // coercion. Issue a raw GET (rather than the typed client, which throws a generic
  // error that hides the status) so the assertion can pin the exact 400.
  const badQuery = await fetch(
    `http://127.0.0.1:${String(port)}/quote/inv-1?minAmount=not-a-number`,
  );
  check(badQuery.status === 400, "malformed query decimal rejected with 400");

  await new Promise<void>((resolve, reject) =>
    server.close((err) => (err ? reject(err) : resolve())),
  );
  console.log("OK");
}

void main();
